kythe/rust/indexer/src/bin/proxy/main.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2020 The Kythe Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+extern crate kythe_rust_indexer;
+use kythe_rust_indexer::{indexer::KytheIndexer, providers::*, writer::ProxyWriter};
+
+use analysis_rust_proto::*;
+use anyhow::{anyhow, Context, Result};
+use serde_json::Value;
+use std::{
+    fs::File,
+    io::{self, Write},
+    path::{Path, PathBuf},
+};
+use tempdir::TempDir;
+
+fn main() -> Result<()> {
+    let mut file_provider = ProxyFileProvider::new();
+    let mut kythe_writer = ProxyWriter::new();
+    let mut indexer = KytheIndexer::new(&mut kythe_writer);
+
+    // Request and process
+    loop {
+        let unit = request_compilation_unit()?;
+        // Retrieve the save_analysis file
+        let temp_dir = TempDir::new("rust_indexer").context("Couldn't create temporary directory")?;
+        let temp_path = PathBuf::new().join(temp_dir.path());
+        write_analysis_to_directory(&unit, &temp_path, &mut file_provider)?;
```

Comment:
So if this fails & returns (or the `TempDir::new`) we don't write "done" back to the proxy, and so the conversation will be out of sync.

Reply:
Gotcha, I'll change the structure of how this all works